crates/codegen/parser/runtime/src/cursor.rs

```diff
@@ -131,35 +121,64 @@ impl Cursor {
 
     /// Returns the currently pointed to [`Node`].
     pub fn node(&self) -> Node {
-        self.current.node.clone()
+        self.node.clone()
     }
 
     pub fn node_name(&self) -> Option<FieldName> {
-        self.path.last().and_then(|parent| {
-            let this = &parent.rule_node.children[self.current.child_number];
+        self.parent.as_ref().and_then(|parent| {
+            let this = &parent.rule_node.children[self.child_number];
 
             this.name
         })
     }
 
     /// Returns the text offset that corresponds to the beginning of the currently pointed to node.
     pub fn text_offset(&self) -> TextIndex {
-        self.current.text_offset
+        self.text_offset
     }
 
     /// Returns the text range that corresponds to the currently pointed to node.
     pub fn text_range(&self) -> TextRange {
-        self.current.text_range()
+        let start = self.text_offset;
+        let end = start + self.node.text_len();
+        start..end
     }
 
     /// Returns the depth of the current node in the CST, i.e. the number of ancestors.
     pub fn depth(&self) -> usize {
-        self.path.len()
+        let mut depth = 0;
```

Comment:
Thoughts about storing the current depth on the cursor, to avoid the traversal every time?

Reply:
I thought that might be premature. Not sure that it would be called that often.